### PR TITLE
Use the initializer to avoid NoMethodError on nil.include?

### PR DESCRIPTION
**Pull request by josb**
Fixes ``` undefined method `include?' for nil:NilClass``` when using tag() in the recipe DSL.

### CODE REVIEW

**Comment by lamont-granquist** [comment]
fix is right, but could use a test.

**Comment by josb** [comment]
A problem with this issue is that recipe_spec.rb does things like
```
Chef::Node.new.tap {|n| n.normal[:tags] = [] }
```
This initializes the object to ```[]```, whereas it would appear that the "regular" code does not (hence the issue I am seeing). So the irony here is that the test framework initialization hides this bug.

**Comment by lamont-granquist** [comment]
node.normal[:tags] is a VividMash and will autovivify to a Mash, so it may be getting initialized on access somewhere.

probably everything should be cleaned to go in through Chef::Node#tags and Chef::Node#tag as the APIs an nothing else should access tags as a hash/mash/attribute directly.

the line in recipe_spec looks pretty wrong and should most likely just be dropped.  looks kind of like someone found the bug and then 'fixed' the test instead of fixing the bug...

**Comment by josb** [comment]
Thanks for the feedback, Lamont.

This should probably be a different issue (i.e. not block this one from landing) but it sounds like we want s/(node|normal)[:tags]/tags/ except for inside the tags method itself? There are quite a few occurrences of the existing access pattern, hence my question.

**Comment by btm** [comment]
@josb in `lib/chef/recipe.rb` the method `Chef::Recipe#tagged?` does `return false if run_context.node[:tags].nil?`, which is correct. We expect methods ending in ? to return boolean values.

`chef$ git grep '\[:tags\]' | grep -v ^spec | grep -v 'config\[:tags\]' | wc` => 11 lines, and 4 of those are the lines inside node that we're talking about saving. That is, I think we should fix tags as @lamont-granquist suggested since it's only a few lines in the code and the associated tests, and see how that changes the behavior so we can come out with a regression test.

**Comment by josb** [comment]
@btm: ```node[:tags]``` being ```nil``` is a bug. We should not have to check for that value, as it's not a valid value for this particular variable. (I'm aware of the Boolean return value convention, btw).

I fixed the other occurrences, please see the current diff. Thanks for your feedback!

File: lib/chef/knife/core/node_presenter.rb [hunk 1]
```diff
@@ -125,7 +125,7 @@ def summarize(data)
             summarized << <<-SUMMARY
 #{key('Recipes:')}     #{Array(node[:recipes]).join(', ')}
 #{key('Platform:')}    #{node[:platform]} #{node[:platform_version]}
-#{key('Tags:')}        #{Array(node[:tags]).join(', ')}
+#{key('Tags:')}        #{node.tags.join(', ')}
```
Comment (btm) [inline lib/chef/knife/core/node_presenter.rb]: We're super sure now that node.tags will always give us an array, because that's part of the bug, right?

Comment (josb) [inline lib/chef/knife/core/node_presenter.rb]: Correct. I fixed all the places where we access node[:tags] which was the only way we'd never initialize it to an Array. Chef::Node#tags makes sure we initialize properly.

**Comment by btm** [comment]
:+1: 

@chef/client-core, review?

**Comment by lamont-granquist** [comment]
I just did a quick skim, :+1: on the overall shape, I didn't get into details.

**Comment by lamont-granquist** [comment]
:+1:  on the details as well

**Comment by thommay** [comment]
:+1: , thanks @josb!

**Comment by lamont-granquist** [comment]
:+1: 

if you want to address Dan's `respond_to?(:read)` comment that'd be cool as well, that's an idiom that I've seen before.

**Comment by josb** [comment]
Thanks, everyone :)

**Pull request merged**